discovery/validation_barrier.go

```diff
@@ -27,3 +58,429 @@ type ValidationBarrier interface {
+	log.Tracef("Signaling dependants for %T, allow=%v", nMsg.msg, allow)
+
+	switch msg := nMsg.msg.(type) {
+	// If we've just finished executing a ChannelAnnouncement, then we'll
+	// close out the signal, and remove the signal from the map of active
+	// ones. This will allow/deny any dependent jobs to continue execution.
+	case *lnwire.ChannelAnnouncement:
+		finSignals, ok := g.chanAnnFinSignal[msg.ShortChannelID]
+
+		// Exit early if there's no dependant.
+		if !ok {
+			return
+		}
+
+		// If this is allow signal, we'll send it to all dependants and
+		// remove it from the map.
+		if allow {
+			close(finSignals.allow)
+			delete(g.chanAnnFinSignal, msg.ShortChannelID)
+
+			return
+		}
+
+		// If this is a deny signal, check if we have any dependants.
+		// If not we'll exit early.
+		denySignal, ok := finSignals.deny[relayingNode]
+		if !ok {
+			return
+		}
+
+		// Otherwise send the deny signal to the specific node and
+		// remove it from the map.
+		//
+		// NOTE: we cannot remove it from `chanAnnFinSignal` yet as
+		// it's shared by other relaying nodes. We will need it here if
+		// a different node is sending signal deny.
+		close(denySignal)
+
+		// TODO(yy): instead of removing the deny signal, we could
+		// count how many deny signals have been sent from this
+		// relaying node and further ban this misbehaving peer.
+		delete(finSignals.deny, relayingNode)
+
+	// For all other job types, we'll delete the tracking entries from the
+	// map if it's a signal allow. Otherwise, we'll do nothing but exit.
+	// The entries will be removed when at least one of the relaying nodes
+	// sends a legitimate ChannelAnnouncement.
+	case *lnwire.NodeAnnouncement:
+		if !allow {
+			return
+		}
+
+		delete(g.nodeAnnDependencies, route.Vertex(msg.NodeID))
+
+	case *lnwire.ChannelUpdate:
+		if !allow {
+			return
+		}
+
+		delete(g.chanEdgeDependencies, msg.ShortChannelID)
+
+	default:
+	}
+}
+
+// getOrCreateNodeSemaphore returns the node's semaphore. It will create the
+// semaphore if it doesn't exist.
+func (g *GossipBarrier) getOrCreateNodeSemaphore(
+	node route.Vertex) chan struct{} {
+
+	g.Lock()
+	defer g.Unlock()
+
+	semaphore, ok := g.nodesSemaphore[node]
+	if ok {
+		return semaphore
+	}
+
+	// Create a buffered channel with `numActiveReqsPerNode` slots.
+	semaphore = make(chan struct{}, numActiveReqsPerNode)
+	g.nodesSemaphore[node] = semaphore
+
+	log.Tracef("Created semaphore for relaying node %v", node)
+
+	return semaphore
+}
+
+// acquireSemaphore reads an item from the globalSemaphore and sends it to
+// the node's semaphore channel. It will block until a slot is accepted by the
+// node semaphore and a slot is returned from the global semaphore. This is
+// equivalent of `nodeSemaphore <- (<-g.globalSemaphore)`.
+func (g *GossipBarrier) acquireSemaphore(node route.Vertex) {
+	nodeSemaphore := g.getOrCreateNodeSemaphore(node)
+
+	// Send to the node's semaphore.
+	select {
+	case nodeSemaphore <- struct{}{}:
+	case <-g.quit:
```

Comment:
I think we want to `return` here instead of attempt to grab the global semaphore. Same below.

Reply:
Good catch! now fixed!